File: programmers/Level1/키패드누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        String answer = "";
        int[][] keypad = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}, {-1, 0, -1}};
        int[] leftPos = {3, 0};        // * 위치에서 시작 -> [3][0]
        int[] rightPos = {3, 2};       // # 위치에서 시작 -> [3][2]
        
        for(int i = 0; i < numbers.length; i++) {
            if(numbers[i] == 1 || numbers[i] == 4 || numbers[i] == 7) {
                answer += "L";
                for(int j = 0; j < 3; j++) {
                    if(keypad[j][0] == numbers[i]) {
                        leftPos[0] = j;
                        leftPos[1] = 0;
                    }
                }
            } else if(numbers[i] == 3 || numbers[i] == 6 || numbers[i] == 9) {
                answer += "R";
                for(int j = 0; j < 3; j++) {
                    if(keypad[j][2] == numbers[i]) {
                        rightPos[0] = j;
                        rightPos[1] = 2;
                    }
                }
            } else {
                int idx = 0;
                for(int j = 0; j < 4; j++) {
                    if(keypad[j][1] == numbers[i]) {
                        idx = j;
                        break;
                    }
                }
                
                int distLeft = Math.abs(leftPos[0] - idx) + Math.abs(leftPos[1] - 1);
                int distRight = Math.abs(rightPos[0] - idx) + Math.abs(rightPos[1] - 1);
                if(distLeft < distRight) {
                    answer += "L";
                    leftPos[0] = idx;
                    leftPos[1] = 1;
                } else if(distLeft > distRight) {
                    answer += "R";
                    rightPos[0] = idx;
                    rightPos[1] = 1;
                } else {
                    if(hand.equals("left")) {
                        answer += "L";
                        leftPos[0] = idx;
                        leftPos[1] = 1;
                    } else {
                        answer += "R";
                        rightPos[0] = idx;
                        rightPos[1] = 1;
                    }
                }
            }
        }
        
        return answer;
    }
}

File: programmers/Level1/키패드누르기.py
def solution(numbers, hand):
    answer = ''
    keypad = [[1, 4, 7, '*'], [2, 5, 8, 0], [3, 6, 9, '#']]
    
    l = [0, 3]
    r = [2, 3]
    
    for num in numbers:
        if num in keypad[0]:
            answer += 'L'
            l = [0, keypad[0].index(num)]
        elif num in keypad[2]:
            answer += 'R'
            r = [2, keypad[2].index(num)]
        else:
            idx = keypad[1].index(num)
            
            dl = abs(l[0] - 1) + abs(l[1] - idx)
            dr = abs(r[0] - 1) + abs(r[1] - idx)
            
            if dl == dr:
                if hand == 'left':
                    answer += 'L'
                    l = [1, idx]
                else:
                    answer += 'R'
                    r = [1, idx]
            elif dl < dr:
                answer += 'L'
                l = [1, idx]
            else:
                answer += 'R'
                r = [1, idx]
                
    return answer